File: frontend/src/api/review.ts
import { API_REVIEW_ENDPOINT, API_STREAM_ENDPOINT } from '@/constants/api';
import type { ReviewData, ReviewDTO, ReviewRequest } from '@/types/review';
import { httpClient } from '@/utils/httpClient';
import { parseReviewResult } from '@/utils/review';

export function mapToReviewData(dto: ReviewDTO): ReviewData {
  return {
    detectedLanguage: dto.detected_language,
    summary: dto.summary,
    severity: dto.severity,
    score: dto.score,
    bugs: dto.bugs,
    securityIssues: dto.security_issues,
    suggestions: dto.suggestions,
    positives: dto.positives,
  };
}

export async function fetchReview(request: ReviewRequest): Promise<ReviewData> {
  const response = await httpClient.post<unknown>(API_REVIEW_ENDPOINT, {
    code: request.code,
    language: request.language.toLowerCase(),
  });
  return mapToReviewData(parseReviewResult(response));
}

export async function streamReview(
  request: ReviewRequest,
  onChunk: (chunk: string) => void,
  signal: AbortSignal
): Promise<void> {
  await httpClient.stream(
    API_STREAM_ENDPOINT,
    { code: request.code, language: request.language.toLowerCase() },
    onChunk,
    signal
  );
}

File: app/routes/review.py
from fastapi import APIRouter, Request
from fastapi.responses import StreamingResponse

from app.models import ReviewRequest, StructuredReviewResponse

router = APIRouter(prefix="/review", tags=["review"])


@router.post("/stream")
async def review_code_stream(request: ReviewRequest, req: Request) -> StreamingResponse:
    return StreamingResponse(
        req.app.state.reviewer.review_stream(request.code, request.language),
        media_type="text/plain",
    )


@router.post("/structured", response_model=StructuredReviewResponse)
async def review_code_structured(
    request: ReviewRequest, req: Request
) -> StructuredReviewResponse:
    result = await req.app.state.reviewer.review_structured(
        code=request.code, language=request.language
    )
    return StructuredReviewResponse(
        detected_language=result.detected_language,
        summary=result.summary,
        severity=result.severity,
        score=result.score,
        bugs=result.bugs,
        security_issues=result.security_issues,
        suggestions=result.suggestions,
        positives=result.positives,
    )
